[Feat] Add latency as an optimization objective

reyna-abhyankar opened the pull request:
Close #20

yiying-zhang commented on cognify/optimizer/evaluator.py:
```diff
@@ -477,29 +486,31 @@ def get_score(
+            nonlocal total_score, total_cost, total_exec_time, n_success
             n_success += 1
-            score, price = eval_result[3], eval_result[4]
+            score, price, exec_time = eval_result[3], eval_result[4], eval_result[6]   # using these indices is dangerous
```
don't use hard-coded indices

jiange91 replied:
Apply weight to different objectives during the search is non-trivial (also not straightforward for users to set), maybe we can add on/off in this PR.

reyna-abhyankar replied:
Done (removed hardcoded indices). Instead, it's an actual data structure. As for allowing users to select which one, I am adding that in a separate PR. This PR will just jointly optimize quality, cost, and latency and report that to the user.

reyna-abhyankar replied:
Btw, see #23 for allowing users to select on/off for different targets (e.g., `["cost", "latency"]` or `["quality", "latency"]`). That one includes the changes made here as well, so if you want to review that one and just merge that, then we can close this. Otherwise, we can merge them 1 at a time.

yiying-zhang reviewed (changes requested):
i'm a bit confused on how exec_time is factored in the optimization flow. Now we have three targets, how do we decide which one is more important (optimize more and spend more optimization resources on). We should allow users to select which goals they want to optimize.

reyna-abhyankar commented:
Merged `patience` from main. Added some small fixes in the case where no solution is found for a single trial, @jiange91 if you can take a look. Also removed hard-coded indices from patience as well
